### ISSUE TITLE
[red-knot] Model implicit imports of submodules?

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#17296]
Consider the following case **([playground](https://playknot.ruff.rs/68b7810b-06b5-4914-b0a2-8bc1d5513181)):**

`main.py`

```py
import module

print(module.submodule.SubmoduleClass)
```

`module/__init__.py`

```py
from module.submodule import SubmoduleClass
```

`module/submodule.py`

```py
class SubmoduleClass: ...
```

This works at runtime due to a quirk in module resolution (the `from module.submodule import SubmoduleClass` import implicitly stores `submodule` as an attribute on `module`, thanks @AlexWaygood).

It is not considered good practice to rely on this, but people will probably write code that relies on it. Neither pyright nor mypy report any errors here. Should we attempt to model this behavior as well?

**Comment by carljm** [comment]
Do mypy/pyright emit errors on the same scenario, but without the submodule import in `__init__.py`?

That is, are they actually modeling the import side effect, or are they simply failing to catch this category of error altogether?

My feeling is that the code as written only works by non-local "accident", and there is only upside to requiring an explicit import of the submodule in the module where it is used. So I like our current behavior.

**Comment by sharkdp** [comment]
*With* the `from module.submodule import SubmoduleClass` import in `module/__init__.py`:

* Mypy: no error
* Pyright: main.py:6:14 - error: "submodule" is not a known attribute of module "module"
* Red Knot: main.py:6:7: Type `<module 'module'>` has no attribute `submodule`

*Without* the `from module.submodule import SubmoduleClass` import in `module/__init__.py`:

* Exactly the same behavior for all type checkers

---

I might have only tried mpyp in the original example. It seems like we are consistent with pyright here. And it seems like `mypy` simply doesn't aim to catch this category of error. So it seems like we can maybe just close this without changing anything?




> EDIT: On further reflection, I think there could be a case for handling this, specifically in the case shown where the module's own `__init__.py` always imports the submodule. This is more robust than the general case, and more plausible for us to detect.

👍 



> The more general case is that this scenario can occur, and work at runtime, with the submodule import located literally anywhere in the codebase, as long as it's in a module that happens to get imported before the submodule access in `main.py` occurs. But this is extremely fragile, as order of imports can easily change unintentionally.

Do not remind me of my C++ times!

> I don't think we should (or can) handle that general case.

Absolutely. In this case, the diagnostic would technically be a "false positive" (it "works" at runtime), but is arguably serving a purpose. Because it allows you to find missing explicit imports that might end up causing trouble.

**Comment by carljm** [comment]
I edited the issue title to reflect (I think) our agreement on what we would change here, if we change anything at all. But I'm comfortable changing nothing at all unless real-world false-positive data convinces us to make this a priority over other things.
